test/unit/@node-red/registry/lib/externalModules_spec.js

```diff
@@ -1,20 +1,302 @@
+    }
+    await fs.ensureDir(homeDir);
+}
 
+async function setupExternalModulesPackage(dependencies) {
+    await fs.ensureDir(path.join(homeDir,"externalModules"))
+    await fs.writeFile(path.join(homeDir,"externalModules","package.json"),`{
+"name": "Node-RED-External-Modules",
+"description": "These modules are automatically installed by Node-RED to use in Function nodes.",
+"version": "1.0.0",
+"private": true,
+"dependencies": ${JSON.stringify(dependencies)}
+}`)
+}
 
 describe("externalModules api", function() {
-    
+    beforeEach(async function() {
+        await createUserDir()
+    })
+    afterEach(async function() {
+        await fs.remove(homeDir);
+    })
+    describe("checkFlowDependencies", function() {
+        beforeEach(function() {
+            sinon.stub(exec,"run", async function(cmd, args, options) {
+                let error;
+                if (args[1] === "moduleNotFound") {
+                    error = new Error();
+                    error.stderr = "E404";
+                } else if (args[1] === "moduleVersionNotFound") {
+                    error = new Error();
+                    error.stderr = "ETARGET";
+                } else if (args[1] === "moduleFail") {
+                    error = new Error();
+                    error.stderr = "Some unexpected install error";
+                }
+                if (error) {
+                    throw error;
+                }
+            })
+        })
+        afterEach(function() {
+            exec.run.restore();
+        })
+        it("does nothing when no types are registered",async function() {
+            externalModules.init({userDir: homeDir});
+            await externalModules.checkFlowDependencies([
+                {type: "function", libs:[{module: "foo"}]}
+            ])
+            exec.run.called.should.be.false();
+        });
+
+        it("skips install for modules already installed", async function() {
+            externalModules.init({userDir: homeDir});
+            externalModules.register("function", "libs");
+            await setupExternalModulesPackage({"foo": "1.2.3", "bar":"2.3.4"});
+            await externalModules.checkFlowDependencies([
+                {type: "function", libs:[{module: "foo"}]}
+            ])
+            exec.run.called.should.be.false();
+        })
+
+        it("skips install for built-in modules", async function() {
+            externalModules.init({userDir: homeDir});
+            externalModules.register("function", "libs");
+            await externalModules.checkFlowDependencies([
+                {type: "function", libs:[{module: "fs"}]}
+            ])
+            exec.run.called.should.be.false();
+        })
+
+        it("installs missing modules", async function() {
+            externalModules.init({userDir: homeDir});
+            externalModules.register("function", "libs");
+            fs.existsSync(path.join(homeDir,"externalModuels")).should.be.false();
```

Comment:
externalModuels is misspelled by accident?

Reply:
Oops, well spotted